src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,18 +358,59 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
-runHook = (hookName, args) ->
+class Hook
+	constructor: (hookFn) ->
+		@hookFn = hookFn
+		@executed = false
+
+	run: (args...) ->
+		Promise.try =>
+			@hookFn(args...)
+		.finally =>
+			@executed = true
+
+class SideEffectHook extends Hook
+	constructor: (hookFn) ->
+		@rollbackFns = []
+		@rolledBack = false
+		super(hookFn)
+
+	registerRollback: (fn) ->
+		runFn = Promise.method(fn)
+		if @rolledBack
+			runFn()
+		else
+			@rollbackFns.push(runFn)
+
+	rollback: ->
+		# set rolledBack to true straight away, so that if any rollback action
+		# is registered after the rollback call, we will immediately execute it
+		@rolledBack = true
+		Promise.each(@rollbackFns, (fn) -> fn())
+
+runHooks = (hookName, args) ->
 	Object.defineProperty args, 'api',
 		get: _.once ->
 			return api[args.request.vocabulary].clone(passthrough: _.pick(args, 'req', 'tx'))
-	hooks = args.req.hooks[hookName] || []
-	hooks = hooks.concat(args.request?.hooks?[hookName] || [])
+	hooks = args.request?.hooks?[hookName] || []
 	Promise.map hooks, (hook) ->
-		hook(args)
+		hook.run(args)
+
+# The execution order of rollback actions is unspecified
+undoHooks = (request) ->
+	Promise.map _.flatten(_.values(request.hooks)), (hook) ->
+		if hook instanceof SideEffectHook and hook.executed
```

Comment:
I don't think there's any need for this `executed` flag - even if the hook hasn't finished we still want to run any rollback hooks it's hit so far, and also enable the instant triggering of any it adds later